### PR TITLE
PR to bring in changes from punkadiddle fork of the PyPI Resource

### PR DESCRIPTION
**Pull request by trulede**
This PR is to bring in the changes from the punkadiddle version of the Concourse PyPI Resource.

I personally have used that version for working with a private PyPI repository, however I've not used it against the public PyPi repo.

### CODE REVIEW

**Comment by mboldt** [comment]
Thank you! Working on this now.

File: pypi_resource/pipio.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        host_login = '%s:%s@%s' % (
+            repocfg.get('username', ''),
+            repocfg.get('password', ''),
+            url_parts[1]
+        )
+        host_login = host_login.lstrip(':@')
+    else:
+        host_login = url_parts[1]
+
+    url = urlunsplit((url_parts[0], host_login, url_parts[2], url_parts[3], url_parts[4]))
```
Comment (mboldt) [inline pypi_resource/pipio.py]: This ends up stripping the trailing slash off of the URL, which causes `twine` to fail the upload:
```
RedirectDetected: "https://test.pypi.org/legacy" attempted to redirect to "https://test.pypi.org/legacy/" during upload. Aborting...
```
On `out`, the `twine` command passes in the user name and password with the command, so are the credentials in the URL necessary in `out`? If not, can we do this url-mangling only when `authenticate: always` and `mode = 'in'`?

**Comment by mboldt** [comment]
Alright, this has now been published at `cfplatformeng/concourse-pypi-resource:latest`. Thank you again!

File: pypi_resource/pipio.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (trulede) [reply pypi_resource/pipio.py]: It seems fairly common to have optional username & password for Concourse resources, so I would probably keep them. I don't exactly see your problem with playing with IDLE, what do you think about this:

```
>>> from urllib.parse import urlsplit
>>> from urllib.parse import urlunsplit
>>> u = urlsplit('https://test.pypi.org/legacy/')
>>> urlunsplit((u[0], 'foo:bar@'+u[1], u[2], u[3], u[4]))
'https://foo:bar@test.pypi.org/legacy/' <<<<======== here the trailing '/' still exists (python 3.7.3)
>>> u.netloc
'test.pypi.org'
```

Still, I think this code can be neater, so I would at the very least do this:

```
url_parts = urlsplit(url)
host_login = ''
if mode == 'in' and repocfg.get('authenticate', None) == 'always':
        host_login = '%s:%s@' % (
            repocfg.get('username', ''),
            repocfg.get('password', ''))
url = urlunsplit((url_parts[0], host_login+url_parts[1], url_parts[2], url_parts[3], url_parts[4]))
return url, url_parts.netloc
```

That has your suggestion in it too ... regarding Twine ... IIRC I had problems there, and eventually found a version of Twine which worked. But still, on my system I don't seem to loose the trailing '/' when looking at this code.

```
>>> u
SplitResult(scheme='https', netloc='test.pypi.org', path='/legacy/', query='', fragment='')
>>> u.path
'/legacy/'
```

Perhaps its taken away somewhere else?

[the remark this answers is not in the document]
Comment (mboldt) [reply pypi_resource/pipio.py]: Yeah, I'm honestly not sure what was going on before, but `twine` is happy now, and everything seems to be working. Thank you again!

**Pull request merged**